src/charm.py

```diff
@@ -1,217 +1,416 @@
+            self._restart_service()
+
+        return is_changed
+
+    def _update_config(self, restart_on_failure: bool = True) -> bool:
+        """Update alertmanager.yml config file to reflect changes in configuration.
+
+        The default alertmanager.yml that is created by alertmanager on startup, if none exist, is:
+
+            route:
+              group_by: ['alertname']
+              group_wait: 30s
+              group_interval: 5m
+              repeat_interval: 1h
+              receiver: 'web.hook'
+            receivers:
+            - name: 'web.hook'
+              webhook_configs:
+              - url: 'http://127.0.0.1:5001/'
+            inhibit_rules:
+              - source_match:
+                  severity: 'critical'
+                target_match:
+                  severity: 'warning'
+                equal: ['alertname', 'dev', 'instance']
+
+        Args:
+          restart_on_failure: a flag indicating if the service should be restarted if a config
+          hot-reload failed.
+
+        Returns:
+          True if anything changed; False otherwise
+        """
+        config: dict = yaml.safe_load(self.container.pull(self._config_path))
+        if pagerduty_key := self.model.config.get("pagerduty_key"):
+            config.update({'receivers': [{
+                'name': 'default_pagerduty',
+                'pagerduty_configs': [{
+                    'send_resolved': True,
+                    'service_key': pagerduty_key}
+                ]}
+            ]})
+            config['route']['receiver'] = 'default_pagerduty'
+        else:
+            # pagerduty_key evaluates to False: restore sections to default values (alertmanager
+            # won't start without a receiver)
+            config.update({'receivers': [{
+                'name': 'web.hook',
+                'webhook_configs': [
+                    {
+                        'url': 'http://127.0.0.1:5001/',
+                    }
+                ]
+            }]})
+            config['route']['receiver'] = 'web.hook'
+
+        config_yaml = yaml.safe_dump(config)
+        config_hash = utils.sha256(config_yaml)
+
+        is_changed = False
+        if config_hash != self._stored.config_hash:
+            is_changed = True
+            # Get a reference to the container so we can manipulate it
+            self.container.push(self._config_path, config_yaml)
+            self._stored.config_hash = config_hash
+
+        if is_changed:
+            # Send an HTTP POST to alertmanager to hot-reload the config.
+            # This reduces down-time compared to restarting the service.
+            api = AlertmanagerAPIClient(self._fetch_private_address(), self._api_port)
+            if not api.reload():
+                logger.warning("config reload via HTTP POST failed")
+                if restart_on_failure:
+                    self._restart_service()
+
+        return is_changed
+
+    def _common_exit_hook(self) -> bool:
```

Comment:
This name is a bit confusing. There seems to be a lot more logic here than just exiting the hook

Reply:
Please do suggest a better name. My intention was not "exiting the hook" but rather a common hook through which everyone is exiting, e.g. `_common_gate`, `_one_hook_to_rule_them_all` etc.

Reply:
I have it as just _configure but it's up to you. I don't even really see this as a problem. I just thought I would mention that one might think this is just exit logic.